sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+    "",       // First property has no parameter
+    32        // Value of second property
+    )]]
+#endif
+  ;
+ public:
+  annotated_ptr(T *p) : ptr(p) {}
+};
+
+} // namespace sycl::ext::oneapi
+```
+
+When the device compiler generates code to reference the decorated member
+variable, it emits a call to the LLVM intrinsic function
+[`@llvm.ptr.annotation`][10] that annotates the pointer to that member
+variables, similar to the way the existing clang `__attribute__((annotate()))`
+works.  Illustrating this with some simplified LLVM IR that matches the example
+code above:
+
+[10]: <https://llvm.org/docs/LangRef.html#llvm-ptr-annotation-intrinsic>
+
+```
+@.str = private unnamed_addr constant [27 x i8] c"sycl-properties:foo,bar=32\00",
+  section "llvm.metadata"
+@.str.1 = private unnamed_addr constant [9 x i8] c"file.cpp\00",
+  section "llvm.metadata"
+
+define void @foo(i32* %ptr) {
+  %aptr = alloca %class.annotated_ptr
+  %ptr = getelementptr inbounds %class.annotated_ptr, %class.annotated_ptr* %aptr,
+    i32 0, i32 0
+  %1 = bitcast i32** %ptr to i8*
+  %2 = call i8* @llvm.ptr.annotation.p0i8(i8* %1,
+    i8* getelementptr inbounds ([27 x i8], [27 x i8]* @.str, i32 0, i32 0),
+    i8* getelementptr inbounds ([9 x i8], [9 x i8]* @.str.1, i32 0, i32 0),
+    i32 3, i8* null)
+  %3 = bitcast i8* %2 to i32**
+  store i32* %ptr, i32** %3
+  ret void
+}
+```
+
+The front-end encodes the properties from the C++ attribute
+`[[__sycl_detail__::add_ir_member_annotation()]]` into the annotation string
+(`@.str` in the example above) using the following algorithm:
+
+* The property value is converted to a string as specified in
+  [Property representation in C++ attributes][5].
+* Construct a property definition string for each property:
+  - If the property value is the empty string, the property definition is just
+    the name of the property.
+  - Otherwise, the property definition string is formed by concatenating the
+    property name with the equal sign (`=`) and the property value.
+* The annotation string is formed by concatenating all property definition
+  strings, separated by a comma (`,`).
+* The annotation string is pre-pended with `"sycl-properties:"` and NULL
+  terminated.
+
+**NOTE**: Calls to the `@llvm.ptr.annotation` intrinsic function are known to
+disable many clang optimizations.  As a result, properties added to a
+non-global variable will likely result in LLVM IR (and SPIR-V) that is not well
+optimized.  This puts more pressure on the SPIR-V consumer (e.g. JIT compiler)
+to perform these optimizations.
+
+
+## Property representation in C++ attributes
+
+As noted above, there are several C++ attributes that convey property names and
+values to the front-end:
+
+* `[[__sycl_detail__::add_ir_global_variable_attributes()]]`
+* `[[__sycl_detail__::add_ir_kernel_parameter_attributes()]]`
+* `[[__sycl_detail__::add_ir_function_attributes()]]`
+* `[[__sycl_detail__::add_ir_member_annotation()]]`
```

Comment:
Could you please clarify why we can't use single attribute in all cases? It seems FE should be able to understand if an attribute is global_variable one or function one(...) by just looking at the entity it is applied to.